Druid automated quickstart

findingrish opened the pull request:
### Description
Setting up a new Druid cluster could involve tuning several runtime properties required by each service. Each process must also be given the appropriate jvm arguments so that the system can perform optimally.

Druid already contains several single-server profiles such as `nano-quickstart`, `micro-quickstart`, `large-quickstart`, etc. But it is still difficult to have a setup which does not match exactly with any of these profiles. A lot of the properties in these profiles can also be commoned out and simplified.

This PR adds a new script `start-druid` in order to:
- simplify starting any combination of Druid services on a single server
- reduce the number of inputs required to start Druid
- deprecate the existing profiles and eventually remove them

### Usage 

#### Command:
```bash
./start-druid [--memory=totalAvailableMemory] [--services=listOfServices] [--config=rootConfigDir]
```

#### Examples:

1. Run all services with total system memory and default runtime properties
`./start-druid`

2. Run one instance each of specific services with given total memory
`./start-druid --services=broker,router,historical --memory=10g`

3. Run services with overridden properties
`./start-druid --services=broker,router,middleManager,historical --config=rootConfigDir`
Directory `_common` must be present inside `rootConfigDir` and must contain files `common.runtime.properties` and `common.jvm.config` with mandatory properties and any other property that you wish to override. Please refer to `conf/druid/auto/_common` for an example.
  - __Override runtime properties of a service__: Specify a `runtime.properties` for that service. For the example command above, create a file  `rootConfigDir/broker/runtime.properties`. This file should contain the properties you wish to override.
  - __Override jvm arguments for a service__:  To override the jvm args for any service, you must override the jvm args for _all_ the services. This is to prevent miscalculations in allocated memory. For the example command above, create directories `broker`, `router`, `middleManager`, `historical` inside the `rootConfigDir`.
Add a file `jvm.config` specifying the following values in each of the service directories.
`- Xmx***`
`- Xms***`
`- XX:MaxDirectMemorySize***` (required only for broker and historical)
For middleManager, alongwith jvm args, you must override `druid.indexer.runner.javaOptsArray` in `rootConfigDir/middleManager/runtime.properties` 

### Design 

#### Flow
- Run bash script `start-druid` with one of the possible invocations
- `start-druid` invokes python script `start-druid-main.py` which computes the memory required for each service
- `start-druid-main.py` invokes `bin/supervise` with the commands for each service to start.
   - example broker command:
   ```
   broker bin/run-druid broker ~/conf/druid/auto '-Xms2392m -Xmx2392m -XX:MaxDirectMemorySize=1595m'
   ```
   - example middle-manager command:
   ```
   !p90 middleManager bin/run-druid middleManager /apache-druid-{{DRUIDVERSION}}/conf/druid/auto '-Xms346m -Xmx346m' '-Ddruid.worker.capacity=5 -Ddruid.indexer.runner.javaOptsArray=["-server","-Duser.timezone=UTC","-Dfile.encoding=UTF-8","-XX:+ExitOnOutOfMemoryError","-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager","-Xms1g","-Xmx1g","-XX:MaxDirectMemorySize=1g"]'
   ```
- `bin/supervise` starts the services, one at a time, by invoking `run-druid` (same as existing flow while using `./bin/start-micro-quickstart` etc.)

#### Parameters
`start-druid` accepts these optional parameters:
| parameter | description | default value |
|-----------|------------|--------------|
| `memory`| total allocated memory for all Druid services and tasks | 80% of total system memory |
| `config`| root config directory, where service-specific `runtime.properties` and `jvm.config` can be given | `examples/conf/druid/auto`|
| `services` | list of services to start | all services (including zookeeper) |
| `zk` | whether to start zookeeper | false |
| `verbose` | print all messages and computed parameter values | false |
| `compute` | do not start Druid, only print computed parameter values | false |

#### Validations
- `memory` should be in valid format (integer followed by m or g)
- `memory` should satisfy the total minimum requirements
- `<rootConfDir>/<service>/jvm.config` should either be present for all the services being started or none
- `<rootConfDir>/middleManager/runtime.properties` must have `druid.indexer.runner.javaOptsArray` parameter, if middleManager is to be started with overridden memory 

#### Memory distribution
| service | total memory ratio (heap + direct) | minimum total memory (MB) |
|--------|------------------------------------|-----------------------------|
| middleManager | 1 | 64 |
| router | 2 | 128 |
| coordinator-overlord | 30 | 256 |
| task | 30 | 1024 |
| broker | 46 | 900 |
| historical | 80 | 900 |

#### Running existing profiles 
```
nano:  4g,  2 cpu 
micro:  16g, 4 cpu
small: 64g, 8 cpu
medium: 128g, 16 cpu
large: 256g, 32 cpu
xlarge: 512g, 64 cpu 
```
For example, starting `small` in verbose mode:
```
> ./start-druid -m=64g --verbose
...
total memory is 65536m
memory used for services & tasks 32768m

Memory distribution for services:
broker, memory_config: -Xms4785m -Xmx4785m -XX:MaxDirectMemorySize=3190m
router, memory_config: -Xms346m -Xmx346m
coordinator-overlord, memory_config: -Xms5201m -Xmx5201m
historical, memory_config: -Xms5548m -Xmx5548m -XX:MaxDirectMemorySize=8322m
middleManager, memory_config: -Xms173m -Xmx173m
middleManager-task, memory_config: ['-Ddruid.worker.capacity=2', '-Ddruid.indexer.runner.javaOptsArray=["-server","-Duser.timezone=UTC","-Dfile.encoding=UTF-8","-XX:+ExitOnOutOfMemoryError","-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager","-Xms1g","-Xmx1g","-XX:MaxDirectMemorySize=1g"]']
```

#### Limitations
- can only launch a single instance of each service
- does not allow overriding the jvm arguments for just one service, either all must be overridden or none

### Major changes
- Added `start-druid`, bash script with functionality as described above 
- Added `start-druid-main`, python script with functionality as described above 
- Added `conf/druid/auto` directory
- Updated `supervise`, to accept service startup commands as arguments, refactored the method which reads the supervise config file 
- Update `run-druid` script to accept optional parameters for jvm argument and middle manager task params 

### A note on runtime properties
The following runtime properties have not been added in the service-specific `runtime.properties` files inside `conf/druid/auto` because their default values computed in the code seem adequate.

| property | service | value computed in Druid code |
|----------|--------|--------------------------|
|druid.server.http.numThreads | all | `max(10, (numCores * 17) / 16 + 2) + 30` |
| druid.processing.numThreads | broker, historical | `max(numCores - 1, 1)` |
| druid.broker.http.numConnections | broker | 20 |
| druid.broker.http.maxQueuedBytes | broker | `max(25m, heap_size * 0.02)` |
| druid.processing.buffer.sizeBytes | broker, historical | `min(1g, directMemory/ (numThreads + numMergeBuffers + 1))` |
| druid.processing.buffer.numMergeBuffers | broker, historical | `max(2, numThreads / 4)` |
| druid.cache.sizeInBytes | | `~(heap_size / 10)` |
| druid.worker.capacity | middle-manager | `max(numCores - 1, 1)` |
| druid.indexer.runner.javaOptsArray | middle-manager | read from `middle-manager/runtime.properties` if specified, otherwise computed by `start-druid` |
| druid.indexer.fork.property.druid.processing.numMergeBuffers | | |
| druid.indexer.fork.property.druid.processing.buffer.sizeBytes | | |
| druid.indexer.fork.property.druid.processing.numThreads | | |
| druid.router.http.numConnections | router | 20 |
| druid.router.http.readTimeout | router | 15 mins |
| druid.router.http.numMaxThreads | router | `max(10, ((numCores * 17) / 16 + 2) + 30)` |

#### Release note
Add a new script `start-druid` which greatly simplifies starting Druid services on a single server.

Usage:
```bash
./start-druid[--memory=totalAvailableMemory] [--services=listOfServices] [--config=rootConfigDir]
```

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

FrankChen021 commented:
This is a great work to ease the initial setup work for users especially when they're new to Druid.

This startup script relies on a Python script, I know that writting python scripts is much simpler than writing bash scripts, but my question is that what if they don't have a Python interpreter? Which version of Python do they need to run this script? Python 2 or Python 3? Will they struggle on installing a python interpreter?

findingrish commented:
> This is a great work to ease the initial setup work for users especially when they're new to Druid.
> 
> This startup script relies on a Python script, I know that writting python scripts is much simpler than writing bash scripts, but my question is that what if they don't have a Python interpreter? Which version of Python do they need to run this script? Python 2 or Python 3? Will they struggle on installing a python interpreter?
 
@FrankChen021 thanks for going over the change. 

> Which version of Python do they need to run this script? 

It requires python 3. 

> Will they struggle on installing a python interpreter?

Installing python is straightforward on most environments, IMO it shouldn't pose any challenge.

gianm reviewed (commented):
Thanks for the contribution! I haven't reviewed the code, but I do have some comments based on building this branch on my machine and running it:

1. On my machine (64GB RAM) the script decided to set total memory = 52GB and then allocated that chunk of memory 100% to on-heap and direct memory for the various processes. It doesn't leave any room for page cache or for tasks on MMs. The calculations should be adjusted to take those items into account. (By "take page cache into account", I mean leave some memory free so the OS can use it to cache mmapped segments.) I suggest comparing the calculated numbers here against the single-server configs we have: https://druid.apache.org/docs/latest/operations/single-server.html. For example, 64GB RAM would match our "small" config, where we give the Historical a `4g` heap; the script currently generates `12752m`.
3. The way the script accepts command-line arguments is nonstandard; consider using `argparse` instead and accepting `--x=y` style arguments. It would help with error messages, too. I tried an invalid `totalMemory` setting and noticed that the error message had a traceback in it, which makes it look messy. `argparse` wouldn't do that.
4. I noticed that when I provide a very low `totalMemory` (like `40m`) the calculations generate negative numbers and then the processes crash. It would be better for the script to print a nice error message.
5. The script should be less chatty in the normal case. Feel free to provide a `--verbose` mode, but normally, I don't think it should print anything. The first line the user should see is the one from `bin/greet`: "Starting Apache Druid."
6. When you have scripts tail-calling out to other scripts, it's best to use `exec` (in bash) and `execv` or one of the other exec variants (in python). This way, each call replaces the original one rather than forming a chain. It is cleaner and uses fewer resources. For the bash script `start-druid`, refer to `post-index-task` for an example of how you can use `exec`. For python refer to https://docs.python.org/3/library/os.html for `os.exec*` functions.

findingrish commented:
> Thanks for the contribution! I haven't reviewed the code, but I do have some comments based on building this branch on my machine and running it:
> 
> 1. On my machine (64GB RAM) the script decided to set total memory = 52GB and then allocated that chunk of memory 100% to on-heap and direct memory for the various processes. It doesn't leave any room for page cache or for tasks on MMs. The calculations should be adjusted to take those items into account. (By "take page cache into account", I mean leave some memory free so the OS can use it to cache mmapped segments.) I suggest comparing the calculated numbers here against the single-server configs we have: https://druid.apache.org/docs/latest/operations/single-server.html. For example, 64GB RAM would match our "small" config, where we give the Historical a `4g` heap; the script currently generates `12752m`.
> 2. The way the script accepts command-line arguments is nonstandard; consider using `argparse` instead and accepting `--x=y` style arguments. It would help with error messages, too. I tried an invalid `totalMemory` setting and noticed that the error message had a traceback in it, which makes it look messy. `argparse` wouldn't do that.
> 3. I noticed that when I provide a very low `totalMemory` (like `40m`) the calculations generate negative numbers and then the processes crash. It would be better for the script to print a nice error message.
> 4. The script should be less chatty in the normal case. Feel free to provide a `--verbose` mode, but normally, I don't think it should print anything. The first line the user should see is the one from `bin/greet`: "Starting Apache Druid."
> 5. When you have scripts tail-calling out to other scripts, it's best to use `exec` (in bash) and `execv` or one of the other exec variants (in python). This way, each call replaces the original one rather than forming a chain. It is cleaner and uses fewer resources. For the bash script `start-druid`, refer to `post-index-task` for an example of how you can use `exec`. For python refer to https://docs.python.org/3/library/os.html for `os.exec*` functions.

@gianm thanks for the feedback.

Regarding `1`, after going over https://druid.apache.org/docs/latest/operations/single-server.html 
I tried to check the ratio (service memory (heap, direct) + task on MMs) / totalMemory 
nano -> (2 + 1)g / 4g 
micro -> (3 + 4)g / 16g
small -> (25 + 6)g / 64g
medium -> (44 + 8)g / 128g
large -> (85 + 16)g / 256g
xlarge -> (99 + 32)g / 512g

except for nano and xlarge profiles, the ratio is close to 50% 
I am thinking of using 50% as default and allow the user to override this ratio

kfaraz commented on examples/bin/druid-quickstart.py:
```diff
@@ -0,0 +1,337 @@
+
+    where options include:
+       totalMemory=<value>
+            Integer value is supported with 'm' or 'g' suffix.
+            Memory for druid services, if totalMemory is not specified
+            80 percent of system memory is used.
+            Note, if service specific jvm config is present,
+            <totalMemory> shouldn't be specified.
+            <totalMemory> should be greater than equals 3g
```
Please try to make the help text a little less verbose. One or two lines for each param should be sufficient.

e.g. for totalMemory

```suggestion
            Total memory for all the processes (services and tasks, if any).
            Default = 80% of total system memory.
            This parameter is ignored if each service already has a jvm.config in the given conf directory.
            e.g. 500m, 4g, 6g
```

findingrish replied:
I have tried to cut it down.

kfaraz commented on examples/bin/druid-quickstart.py:
```diff
@@ -0,0 +1,337 @@
+       computeOnly
+            Validate the arguments and display memory distribution for services.
+
+    sample usage:
```
List out the example usages first, then explain the parameters in detail. The examples should be self-explanatory, so I don't think we need a description for each. If we do include a description, it should not be more than one line.

findingrish replied:
`argparse` lib is generating the help documentation now.

kfaraz reviewed (commented):
Done a partial review. Added some comments.

findingrish commented on examples/bin/druid-quickstart.py:
```diff
@@ -0,0 +1,337 @@
+
+QUICKSTART_BASE_CONFIG_PATH = "conf/druid/single-server/quickstart"
+HELP_ARG_IDENTIFIER = "help"
+COMPUTE_ONLY_ARG_IDENTIFIER = "computeOnly"
```
These constants have been removed. Using simpler names for existing constants

findingrish commented on examples/bin/druid-quickstart.py:
```diff
@@ -0,0 +1,337 @@
+import sys
```
Added.

findingrish commented on examples/bin/run-druid:
```diff
@@ -47,7 +47,35 @@ if [ ! -d "$LOG_DIR" ]; then mkdir -p $LOG_DIR; fi
+
+if [ "$#" -eq 3 ]
+then
+
```
This is outdated, passing extra params for MM task properties to `run-druid`

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+    MM
+]
+
+SERVICE_MEMORY_DISTRIBUTION_WEIGHT = {
```
Nit: Perhaps `SERVICE_MEMORY_RATIO` would be better?

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+sample usage:
+    start-druid
+            Start up all the services (including zk).
+            50 - 80 percent of system memory is used.
```
This seems vague and a little confusing, maybe just omit this line as you are already exposing the `--verbose` and the `--compute` options.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+            Start up all the services (including zk).
+            50 - 80 percent of system memory is used.
+    start-druid -m=100g
+            Start up all the services (including zk)
+            using the given memory.
```
```suggestion
            Starts all Druid services (including zk) using a total memory of 100GB.
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+            Start up all the services (including zk)
+            using the given memory.
+    start-druid -m=100g --compute_only
+            Compute memory distribution and validate
```
```suggestion
            Does not start Druid, only displays the memory allocated to each service if started with the given total memory.
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+            Compute memory distribution and validate
+            arguments.
+    start-druid -m=100g -sl=broker,router
+            Start broker & router service, using 100g of memory.
```
```suggestion
            Starts a broker and a router, using a total memory of 100GB.
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+            Read config from conf/druid/single-server/quickstart.
+    start-druid -m=100g --sl=broker,router \\
+    -cp=conf/druid/single-server/custom
+            Start broker & router service, using 100g of memory.
+            Read config from <config_path>.
+            Since <memory> is specified, exception is thrown if
+            jvm.config is present for any of the services.
```
```suggestion
            Starts a broker and a router, using a total memory of 100GB.
            Reads configs from the given config path.
            Throws an exception if there is a jvm.config present in any of the service directories inside the given root conf directory.
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+            If jvm.config is specified for both the
+            services within <config_path>/<service>,
+            memory distribution is not calculated.
+            If jvm.config is present for either of the services,
+            exception is thrown.
```
```suggestion
            Throws an exception if jvm.config is present in only some of the service directories.
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+            jvm.config is present for any of the services.
+    start-druid -sl=broker,router \\
+    -cp=conf/druid/single-server/custom
+            Start broker & router service, using system memory.
+            If jvm.config is specified for both the
+            services within <config_path>/<service>,
+            memory distribution is not calculated.
```
```suggestion
            Starts a broker and a router service, using upto 80% of the total system memory.
            Calculates the memory requirements for each service if no jvm.config is present for any of the services.
            Does not calculate memory requirements if jvm.config is present in all of the service directories.
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+                             '{broker, router, middleManager, historical, coordinator-overlord}. \n'
+                             'If the argument is not given, all services \n'
+                             'and zookeeper is started. e.g. -sl=broker,historical')
+    parser.add_argument('--config_path', '-cp', type=str, required=False,
```
Nit: Maybe rename this param to `--configs` or `--config`

findingrish replied:
I am using `--config`.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+
+    if jvm_config_count > 0 and (jvm_config_count != len(service_list) or total_memory != ""):
+        if jvm_config_count != len(service_list):
+            error_and_exit("jvm.config file should be present for all services or none")
```
I think it would be better to raise an exception here (and in other such cases) rather than simply exiting. You can catch these exceptions where you are already catching the KeyboardInterrupt.

findingrish replied:
Done. This looks better.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+        return lower_bound_memory
+
+def build_mm_task_javaopts_array(memory_type):
+    task_memory = f'{MM_TASK_JAVAOPTS_PROP}=['
```
Rename to task_java_opts.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+
+    service_list = services.copy()
+    if MM in services:
+        service_list.append(MM_TASK)
```
shouldn't this take into account the number of tasks?

findingrish replied:
Individual task memory is fixed, either 512m or 2g (heap + direct memory). 

This part of the code computes the total memory for all tasks and using that it computes the number of tasks.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+    MM_TASK: 15
+}
+
+SERVICE_MEMORY_LOWER_BOUND = {
```
```suggestion
MINIMUM_MEMORY_MB = {
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+    COORDINATOR: 15,
+    BROKER: 23,
+    HISTORICAL: 40,
+    MM_TASK: 15
```
Is this weight for a single task or all tasks put together?

If it's for all tasks, having a constant value for this might not be the best. Larger configurations would have several tasks whereas smaller configurations would have fewer tasks. Thus the weights would change.

If it's for a single task, the function `distribute_memory_over_services` should take into account the number of tasks.

findingrish replied:
> Is this weight for a single task or all tasks put together? 

The weight is for all the tasks put together. 

> If it's for all tasks, having a constant value for this might not be the best. Larger configurations would have several tasks whereas smaller configurations would have fewer tasks. Thus the weights would change.

The memory allocated for all tasks would increase proportionately for larger configuration. Since task memory is fixed, it would automatically lead to higher task count. 
I didn't get the part about weights changing? do you mean to say we should allocated bigger chunk of memory in larger profiles?

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+            if service in allocated_services:
+                continue
+            memory_weight_sum += SERVICE_MEMORY_DISTRIBUTION_WEIGHT.get(service)
+        multiplier = (total_memory - lower_bound_memory_allocation) / memory_weight_sum
```
If the input memory was very low, can this become negative?

findingrish replied:
No, there is already a check in `check_memory_constraint` method to ensure that the given memory is greater 
than sum of lower bound memory for the services

```
def check_memory_constraint(total_memory, service_list):
    # 80% of total memory >= sum of lower bound service memory should be
    lower_bound_memory = 0

    for service in service_list:
        lower_bound_memory += MINIMUM_MEMORY_MB.get(service)

    required_memory = int(lower_bound_memory / 0.8)

    if total_memory < required_memory:
        raise ValueError(f'Minimum memory required for starting services is {required_memory}m')

    if total_memory >= 2 * lower_bound_memory:
        return int(total_memory / 2)
    else:
        return lower_bound_memory
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+        allocated_memory = SERVICE_MEMORY_DISTRIBUTION_WEIGHT.get(service) * multiplier
+        if service in SERVICE_MEMORY_LOWER_BOUND and allocated_memory < SERVICE_MEMORY_LOWER_BOUND.get(service):
+            allocated_memory = SERVICE_MEMORY_LOWER_BOUND.get(service)
+            service_memory_config[service], allocated_memory = build_memory_config_string(service, allocated_memory)
```
Why is allocated_memory being reassigned here?

findingrish replied:
Memory allocated to task is multiple of either 512MB or 2GB.
So the reassignment is basically this computation:
`allocated_memory = allocated_memory - task_count * task_memory`.

kfaraz replied:
Makes sense.

Style suggestion:
If this is going to happen only for tasks, it would be better to have a separate logic for tasks in this function itself rather than expecting the common function `build_memory_config_string` to do the right thing. You can maybe break up the `build_memory_config_string` into two methods.

For both cases, it would be cleaner to have the memory (and task count) computed and returned by that function and construct the config string here itself. (This is to avoid having a name such as `build_memory_config_string` which is actually also updating the allocated memory and other params)

findingrish commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+                        help='Total memory for all processes (services and tasks, if any). \n'
+                             'This parameter is ignored if each service already has a jvm.config \n'
+                             'in the given conf directory. e.g. 500m, 4g, 6g\n')
+    parser.add_argument('--service_list', '-sl', type=str, required=False,
```
I am using the name `services` now, I had tried using hyphen and was running into some issues.

findingrish commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+]
+
+SERVICE_MEMORY_DISTRIBUTION_WEIGHT = {
+    MM: 0.5,
```
I have made the changes, I had kept it this way since it was easier to find what percentage is going to each service.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,436 @@
+#!/usr/bin/env python3
```
I guess it's better to retain the `.py` so that the intent is clear. This script is not invoked directly anyway.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,429 @@
+MINIMUM_MEMORY_MB = {
+    MIDDLE_MANAGER: 64,
+    ROUTER: 128,
+    TASK: 1024,
```
Isn't the mimum for task 256mb for lower configs? Is that handled separately?

findingrish replied:
The nano profile support running 2 tasks for 512m (256m heap + 256m direct) each 
Thus the lower bound of 1g.

kfaraz replied:
Okay, so this limit is for all tasks put together. Maybe rename the constant `TASK`  accordingly then.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,429 @@
+    -c=conf/druid/single-server/custom
+            Starts a broker and a router, using a total memory of 100GB.
+            Reads configs from the given config path.
+            Throws an exception if there is a jvm.config present in any
```
On second thought, I think we can get rid of this detail too.
If the user gives incorrect inputs, we would anyway print the right error message. So we need not describe the possible exceptions in the sample usage.

findingrish replied:
The idea of mentioning these error states was to let the users know before starting the script itself, otherwise they would have to try out the script multiple time to get the config right?

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,429 @@
+            Compute memory distribution and validate arguments.
+    start-druid -m=100g -s=broker,router
+            Starts a broker and a router, using a total memory of 100GB.
+            Reads config from conf/druid/single-server/quickstart.
```
Isn't this true for the above commands as well? Maybe just leave out this part.

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,429 @@
+    start-druid -m=100g --s=broker,router \\
+    -c=conf/druid/single-server/custom
+            Starts a broker and a router, using a total memory of 100GB.
+            Reads configs from the given config path.
```
```suggestion
            Reads configs for each service (jvm.config, runtime.properties) from respective folders inside the given root config path.
```

kfaraz commented on examples/bin/start-druid-main:
```diff
@@ -0,0 +1,429 @@
+            of the service directories inside the given root conf directory.
+    start-druid -s=broker,router \\
+    -c=conf/druid/single-server/custom
+            Starts a broker and a router service, using upto 80% of the
+            total system memory. Calculates the memory requirements for
+            each service if no jvm.config is present for any of the
+            services. Does not calculate memory requirements if jvm.config
+            is present in all of the service directories. Throws an exception
+            if jvm.config is present in only some of the service directories.
```
We can probably leave out the extra details here as well:

You can rephrase it like (or something better that you prefer)
```suggestion
            Starts a broker and a router service, reading service configs from the given root directory.
            Calculates memory requirements for each service, if required, using upto 80% of the total system memory.
```

abhishekagarwal87 commented:
@findingrish - Please also verify that docker can use this script.

kfaraz reviewed (approved):
LGTM. +1 after CI passes.

kfaraz commented:
@findingrish , couple of minor observations on running the `start-druid`
- `start-druid` should be made an executable file if we want to users to invoke the file directly
- the `supervise` should be passed `:notify bin/greet` same as the existing profiles do

kfaraz commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,545 @@
+TASK_MEM_TYPE_LOW = "low"
+TASK_MEM_TYPE_HIGH = "high"
+TASK_MEM_MAP = {
+    TASK_MEM_TYPE_LOW: ["-Xms256m", "-Xmx256m", "-XX:MaxDirectMemorySize=256g"],
```
```suggestion
    TASK_MEM_TYPE_LOW: ["-Xms256m", "-Xmx256m", "-XX:MaxDirectMemorySize=256m"],
```

findingrish commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,545 @@
+    java_opts_list = TASK_JAVA_OPTS_ARRAY + mem_array
+
+    for item in java_opts_list:
+        task_memory += f'\"{item}\";'
```
Args to supervise script is being passed with comma as the separator, thus we are using semicolon here. supervise script is converting the semicolon to comma before passing it to middleManager.

kfaraz replied:
Should it maybe be the other way around?
Commands separated by semi-colon and items in array separated by comma?

findingrish replied:
I tried that out, since it saves extra the array transformation, but the perl script seemed to be working with comma separated args only.

kfaraz replied:
Okay, thanks for checking.

kfaraz replied:
Yeah, I guess the `;` would prematurely terminate the command. You could try putting the whole thing inside quotes but that would require you to correctly escape the quotes which occur inside.

findingrish replied:
That was the very first thing I tried, but the perl script wasn't able to parse the args correctly.

findingrish commented:
> @findingrish , couple of minor observations on running the `start-druid`
> 
> * `start-druid` should be made an executable file if we want users to invoke the file directly

It is an executable file, did you run into any issues running it? 

> * the `supervise` should be passed `:notify bin/greet` same as the existing profiles do

Yes, greet command is being passed. Check https://github.com/apache/druid/pull/13365/files#diff-3c1412149c32aa6dd99c6fc15aa69c6f33bba8f31510e1e0f1074e8c69ef5d94R465

kfaraz commented on examples/bin/supervise:
```diff
@@ -179,13 +195,28 @@ usage() unless GetOptions(
   'chdir=s',
-  'svlogd:s'
+  'svlogd:s',
+  'array|a=s{,}'
```
Nit: maybe rename this to something like `commands`

findingrish replied:
`commands` is already a declared var in the script. I can rename it to `command_array`

kfaraz commented:
Thanks, @findingrish , rebuilding fixed the issues.

gianm reviewed (commented):
Shaping up pretty well! Other than the line comments, one high level comment. Please update the docs for this; in particular:

- update the tutorials to use `start-druid` instead of `start-micro-quickstart`.
- update `operations/single-server.md` to mention this script in addition to the other ones.

findingrish commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,545 @@
+    # 5. lastly, if middleManager is to be started, and it is missing task memory properties
+    if jvm_config_count > 0 or mm_task_property_present:
+        if total_memory != "":
+            raise ValueError("If jvm.config is given for services, memory argument shouldn't be specified")
```
The error statement is inaccurate. 
I can modify it to say `If jvm.config for services and/or middleManager configs (druid.worker.capacity, druid.indexer.runner.javaOptsArray) is given, memory argument shouldn't be specified`

findingrish commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,545 @@
+    return task_memory
+
+
+def compute_tasks_memory(allocated_memory):
```
The reason for it being nonmonotonic, is that the task memory changes. 
If total memory for tasks < 2048m, task memory is 512m (256 heap + 256 direct) 
If total memory >= 2048m, task memory is 1g (512m heap + 512m direct) 

 I will update the logic to ensure task count > 2 and is monotonic.

vogievetsky commented on docs/operations/single-server.md:
```diff
@@ -42,37 +43,43 @@ The startup scripts for these example configurations run a single ZK instance al
-- Launch command: `bin/start-nano-quickstart`
-- Configuration directory: `conf/druid/single-server/nano-quickstart`
+- Launch command: `bin/start-druid --memory=4g`
+- Configuration directory: `conf/druid/single-server/quickstart`
```
do we still need to list out the `Configuration directory:` is it is the same in every instance?

kfaraz replied:
No, we shouldn't have to specify the configuration directory as it is the same for all these profiles.

findingrish replied:
I adhered to the existing format, but this is redundant information. I will remove it.

vogievetsky commented on docs/operations/single-server.md:
```diff
@@ -42,37 +43,43 @@ The startup scripts for these example configurations run a single ZK instance al
 
+The `start-druid` is a generic launch script for starting druid services on single server, it accepts optional arguments like services, memory and config.
+All reference configurations can be acheived by passing appropriate arguments to this script.
+Existing launch scripts are deprecated and will be removed in the next release. 
```
This should have a better word than "Existing", maybe "Other"? After all when someone reads this the `bin/start-druid` will exist also

kfaraz replied:
+1

gianm replied:
> Existing launch scripts are deprecated and will be removed in the next release.

I'd rather not include this line ☝️ at all, because IMO we should make the decision to drop the old scripts separately from adding this new feature. Also, never promise anything will happen in "the next release". It's almost a guarantee that it won't happen 🙂

vogievetsky commented on docs/tutorials/index.md:
````diff
@@ -72,38 +73,39 @@ The distribution directory contains `LICENSE` and `NOTICE` files and subdirector
 This brings up instances of ZooKeeper and the Druid services:
 
 ```bash
 $ ./bin/start-micro-quickstart
````
should this command change?

gianm replied:
`./bin/start-druid`

vogievetsky commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,548 @@
+#!/usr/bin/env python3
```
is this chill? should there be a p2 + p3 version? Have a look at how `post-index-task` does it - it is a bash script that routes to the appropriate python because we can not assume that the user has python3

findingrish replied:
I haven't tried out running with p2, since it was end of life. I will update so that it works with both p2 and p3.

gianm replied:
Ah, even if we rely on Python 3, there's still another issue: sometimes people don't literally have `python3` on their path, even though python 3 is available. It's sometimes only called `python`. So this should do the same launcher trick that we have in `examples/bin/post-index-task`, where we try both.

If it's easy to support python 2 _and_ 3, please do that. It will make life easier for users on older OSes.

Either way, we should also add a system requirement for whichever Python versions we end up supporting. Interpreters aren't always present and people may need to install them. Please add this in the places that we mention system requirements:

- after "Java 8u92+" in `tutorials/index.md`
- the "Select OS" section in `tutorials/cluster.md`

As @FrankChen021 mentioned, users might also not be sure how to install a Python interpreter if they don't have one already. So it would also be helpful to describe how to install them on common OSes. For example, on Ubuntu, it can be installed by running:

```
apt-get -y update
apt-get -y install python3
```

Btw, one reason we chose perl for the `bin/supervise` script is that perl is more widely built-in to OSes than python. I guess a lot of OSes use it for their internal scripting. This is also why we don't list perl in the system requirements: it seems to really be ubiquitous. IMO, it's still ok to use python, but we should take these steps to make it easy for people to satisfy the requirement.

findingrish replied:
Added support for python2 and python3.
Updated system requirements. 
Added steps for installing Python interpreter.

vogievetsky commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,548 @@
+                             'in the given conf directory. e.g. 500m, 4g, 6g\n')
+    parser.add_argument('--services', '-s', type=str, required=False,
+                        help='List of services to be started, subset of \n'
+                             '{broker, router, middleManager, historical, coordinator-overlord}. \n'
```
what about `indexer`?

vogievetsky replied:
can we run `coordinator` and `overlord` separately or are we officially removing that option?

kfaraz replied:
`indexer` is not supported yet since we had started based on the `single-server` profiles. We can add support for this in a follow up PR. 
As for `coordinator/overlord`, both `cluster` and `single-server` confs seem to be launching them as one, but we should support launching them separately, maybe in a later PR.

findingrish replied:
> what about indexer?

Running `indexer` isn't supported at the moment, reason being that I was targeting compatibility with the single-server launch scripts and they don't run indexer. Let me see if I can support running indexer. 

> can we run coordinator and overlord separately or are we officially removing that option?

I haven't given it a try, reason being same as above. I will check and get back on this.

gianm replied:
IMO it's ok to omit support for these items in this initial version, since the initial target is to replicate what the prewritten configs do. We can add these items later if needed.

vogievetsky replied:
I guess I was asking in the context of my flow right now for my local Druid is:

- build Druid
- apply some changes to the micro-quickstart configs (switch from MM to indexer for my local testing)
- run `./bin/start-micro-quickstart`

I was wondering how I could update that workflow once `./bin/start-druid` becomes a thing. Can I get `./bin/start-druid` to write out the configs for me? I want to use the awesome memory tuning features but I also want to use the indexer locally. What do?

findingrish replied:
> I was wondering how I could update that workflow once ./bin/start-druid becomes a thing. Can I get ./bin/start-druid to write out the configs for me? I want to use the awesome memory tuning features but I also want to use the indexer locally. What do?

This script doesn't recognise `indexer` service as of now. So I think we will have to use one of the hard-coded profiles.

vogievetsky commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,548 @@
+import multiprocessing
+import argparse
+
+QUICKSTART_ROOT_CONFIG_PATH = "conf/druid/single-server/quickstart"
```
If I understand this correctly these configs will be used as a base and then they will be "extended" when the server starts using the info from the `--memory` config. If so I think we should not call these configs "quickstart". To me "quickstart" means the shitty cluster you spin up on your laptop just to check stuff out (vs the "real" cluster you will deploy later). If my understanding is correct I would call these "base" configs or something similar because you might actually go into production with your start script being `bin/start-druid -m 64g` or something light that.

Also this is not really single server right? You might run `bin/start-druid -s broker,router` on once server and then `bin/start-druid -s historical,middleManager` and `bin/start-druid -s coordinator-overlord` on other servers to make a multi-server cluster.

I would have the configs live in a new location like `conf/druid/base` or `conf/druid/dynamic`

kfaraz replied:
Makes sense. We might eventually get rid of `single-server` and `cluster` directories anyway, since the new script would become the single entry point for everything.

findingrish replied:
> If I understand this correctly these configs will be used as a base and then they will be "extended" when the server starts using the info from the --memory config.

yes, that is correct. 

> If my understanding is correct I would call these "base" configs or something similar because you might actually go into production with your start script being bin/start-druid -m 64g or something light that.

yes, I think `base` would be a better name. 

> Also this is not really single server right? You might run bin/start-druid -s broker,router on once server and then bin/start-druid -s historical,middleManager and bin/start-druid -s coordinator-overlord on other servers to make a multi-server cluster.

yes, that is correct. 

In the very first commit I had a new directory under `conf/druid` with a similar name, it was moved under `single-server` since we were targeting compatibility with single-server launch scripts. But I think it will be better, if these configs live under `conf/druid/base`.

gianm replied:
Makes sense to me too, in which case we can alter my suggestions for the `single-server.md` page; instead of doing `conf/druid/single-server/auto` it would be `conf/druid/base`.

Or maybe `start-druid-base`, so it can be on-its-face obvious that it's connected to the `bin/start-druid` script?

I am ok with a variety of options here, so as far as I am concerned, feel free to do what you think makes the most sense given what we have discussed.

findingrish replied:
I am calling the script `start-druid-auto` and the config directory is `conf/druid/auto`. 

auto (or dynamic) was conveying the capabilities of the script better than `base`. 
Since `auto` was short I picked it over `dynamic`.

vogievetsky commented on docs/operations/single-server.md:
```diff
@@ -42,37 +43,43 @@ The startup scripts for these example configurations run a single ZK instance al
+- Launch command: `bin/start-druid --memory=512g`
+- Configuration directory: `conf/druid/single-server/quickstart`
 
+Memory argument (`--memory`) in the above launch command is optional, if not specified Druid will use upto 80% of system memory.
```
I feel like this is a really key point! It is really buried here. I missed it after reading the docs twice and discovered this by playing and running `bin/start-druid --verbose` do we even need to list all the configurations above?

kfaraz replied:
I think we need a separate section that explains the usage of `start-druid` and the calls out the key points. It can mostly be taken from the description of this PR. @findingrish , maybe you could add a header right before `Single server reference configurations` and briefly explain the usage and key points such as this point about the memory.

findingrish replied:
I will try to place it right where we introduce the `start-druid` script. 

> do we even need to list all the configurations above?

I feel it might help the users choose a machine for trying things out.

gianm replied:
> I feel it might help the users choose a machine for trying things out.

This is a good point! It is helpful to anchor people on a few specific sizes in case they are picking a VM rather than their laptop.

kfaraz commented on docs/operations/single-server.md:
```diff
@@ -42,37 +43,43 @@ The startup scripts for these example configurations run a single ZK instance al
 
 The example configurations run the Druid Coordinator and Overlord together in a single process using the optional configuration `druid.coordinator.asOverlord.enabled=true`, described in the [Coordinator configuration documentation](../configuration/index.md#coordinator-operation).
 
+The `start-druid` is a generic launch script for starting druid services on single server, it accepts optional arguments like services, memory and config.
+All reference configurations can be acheived by passing appropriate arguments to this script.
+Existing launch scripts are deprecated and will be removed in the next release. 
```
```suggestion
`start-druid` is a generic launch script capable of starting any set of Druid services on a server.
It accepts optional arguments such as list of services, total memory and a config directory to override default jvm arguments and service-specific runtime properties.
All other reference configurations (e.g. `micro`, `small`, `xlarge`) can be obtained by passing the
correct memory to this script. The corresponding launch scripts (e.g. `start-micro-quickstart`) are
now deprecated and will be removed in subsequent releases. 
```

gianm commented on docs/operations/single-server.md:
```diff
@@ -42,37 +43,43 @@ The startup scripts for these example configurations run a single ZK instance al
-### Nano-Quickstart: 1 CPU, 4GiB RAM
+### Nano: 1 CPU, 4GiB RAM
 
-- Launch command: `bin/start-nano-quickstart`
```
Please revert these changes -- this part of the doc is meant to describe the nano, micro, etc, bundled configs, which have their own run commands and config directories. We may change that at some point but we haven't yet.

gianm commented on docs/tutorials/index.md:
```diff
@@ -23,7 +23,7 @@ title: "Quickstart (local)"
 
 
-This quickstart gets you started with Apache Druid using the [`micro-quickstart`](../operations/single-server.md#micro-quickstart-4-cpu-16gib-ram) configuration, and introduces you to Druid ingestion and query features.
+This quickstart gets you started with Apache Druid using the [`micro`](../operations/single-server.md#micro-4-cpu-16gib-ram) configuration, and introduces you to Druid ingestion and query features.
```
Ah, I meant to suggest updating the tutorial to use `bin/start-druid` rather than the micro config. I don't think we need to mention the size of the machine at all here, since it's automatic now. Something like this is good:

```
This quickstart gets you started with Apache Druid and introduces you to Druid ingestion and query features.
```

gianm commented on docs/tutorials/index.md:
```diff
@@ -37,7 +37,8 @@ Druid supports a variety of ingestion options. Once you're done with this tutori
 
-Druid comes equipped with several [startup configuration profiles](../operations/single-server.md) for a
+Druid comes equipped with a single launch script that can be used to run several 
+[startup configuration profiles](../operations/single-server.md) for a
```
Here, I'd mention the automatic one first first. Something like:

> Druid comes equipped with launch scripts that can be used to start all processes on a single server. Here, we will use `bin/start-druid`, which automatically sets various runtime properties based on available processors and memory.
>
> In addition, Druid includes several `[bundled non-automatic profiles](../operations/single-server.md)` for a range of machine sizes. These range from `nano` (1 CPU, 4GiB RAM) to `x-large` (64 CPU, 512GiB RAM). We won't use those here, but for more information, see `[Single server deployment](../operations/single-server.md)`. For additional information on deploying Druid services across clustered machines, see `[Clustered deployment](./cluster.md)`.

gianm commented on docs/tutorials/index.md:
````diff
@@ -72,38 +73,39 @@ The distribution directory contains `LICENSE` and `NOTICE` files and subdirector
 
 ```bash
-./bin/start-micro-quickstart
+./bin/start-druid --memory=16g
````
Let's remove `--memory=16g` and allow it to be automatically sized.

gianm commented on docs/tutorials/index.md:
```diff
@@ -72,38 +73,39 @@ The distribution directory contains `LICENSE` and `NOTICE` files and subdirector
 ## Start up Druid services
 
-Start up Druid services using the `micro-quickstart` single-machine configuration.
+Start up Druid services using the `micro` single-machine configuration.
```
`auto`, not `micro`

gianm replied:
(or whatever else we determine based on the other discussion we're having)

gianm commented on docs/tutorials/tutorial-kafka.md:
```diff
@@ -30,7 +30,7 @@ The tutorial guides you through the steps to load sample nested clickstream data
 ## Prerequisites
 
-Before you follow the steps in this tutorial, download Druid as described in the [quickstart](index.md) using the [micro-quickstart](../operations/single-server.md#micro-quickstart-4-cpu-16gib-ram) single-machine configuration and have it running on your local machine. You don't need to have loaded any data.
+Before you follow the steps in this tutorial, download Druid as described in the [quickstart](index.md) using the [micro](../operations/single-server.md#micro-4-cpu-16gib-ram) single-machine configuration and have it running on your local machine. You don't need to have loaded any data.
```
`auto`, not `micro`

gianm commented on docs/tutorials/tutorial-batch-hadoop.md:
```diff
@@ -28,7 +28,7 @@ This tutorial shows you how to load data files into Apache Druid using a remote
 For this tutorial, we'll assume that you've already completed the previous
 [batch ingestion tutorial](tutorial-batch.md) using Druid's native batch ingestion system and are using the
-`micro-quickstart` single-machine configuration as described in the [quickstart](index.md).
+`micro` single-machine configuration as described in the [quickstart](../operations/single-server.md#micro-4-cpu-16gib-ram).
```
`auto`, not `micro`

findingrish commented on examples/bin/supervise:
```diff
@@ -179,13 +195,28 @@ usage() unless GetOptions(
   'chdir=s',
-  'svlogd:s'
+  'svlogd:s',
+  'array|a=s{,}'
```
I have a question regarding your suggestion. How will this line `@config_lines = @{${'commands'}}` work given `commands` is not declared? 

Did you mean something like this?

```
usage() unless GetOptions(
  \%opt,
  'conf|c=s',
  'vardir|d=s',
  'kill-timeout|t=i',
  'chdir=s',
  'svlogd:s',
  'command=s@' => \@commands
);

usage() unless ((@commands || $opt{'conf'}) && $opt{'vardir'});

my @config_lines;

# get commands to execute either from reading the config file or command line
if (not defined $opt{'conf'}) {
    @config_lines = @commands
} else {
    @config_lines = read_config_file($opt{'conf'});
}
```

gianm replied:
I don't think there is a need to declare a target for each argument since `\%opt` is the first arg here. GetOptions has two modes:

1. Provide a hashref (like `\%opt`) first; all options get written as keys to that hashref.
2. Provide a target for each option.

We're using (1) here.

findingrish replied:
It is working for me now, I was trying to access it like `$opt{'command'}` instead of `@{$opt{'command'}}`.

findingrish commented:
Thanks for the review @gianm, @vogievetsky & @kfaraz 

Summarising the changes I have made:
* Added support for Python2 
* Added a new entry point script called `start-druid-auto`, renamed the python script to `start-druid-main` 
* Moved the configs under `conf/druid/auto` 
* Changes in supervise script to read repeated arguments, removed splitting logic, java_opts array transformation 
* Added brief description about the new script in operations/single-server.md, added `auto` profile 
* Rest of the docs refer `auto` profile instead of `micro-quickstart` 
* Added installation instructions for python interpreter

gianm commented on docs/operations/python.md:
```diff
@@ -0,0 +1,46 @@
+  ~ under the License.
+  -->
+
+Apache Druid startup script requires a python interpreter.
```
Which version of interpreter is required?

findingrish replied:
Updated. It works with both Python2 and Python3 interpreter.

gianm commented on docs/operations/python.md:
```diff
@@ -0,0 +1,46 @@
+
+### Linux
+
+#### Debian
```
"Debian or Ubuntu" would be clearer, in case the reader does not know that Ubuntu is based on Debian.

gianm commented on docs/operations/python.md:
```diff
@@ -0,0 +1,46 @@
+## Python interpreter installation instructions
+
+### MacOS
+MacOS comes with a version of Python2, which has been deprecated. Python2 and Python3 can coexist on the same machine without problems.
```
I don't think this is accurate anymore; see https://www.jamf.com/blog/python2-is-gone-apple-macos/. The python situation on macOS is evolving. We should do our best to describe how to get a working interpreter on the 3–4 most recent major macOS releases, which will involve different instructions for different macOS releases.

gianm commented on docs/operations/python.md:
```diff
@@ -0,0 +1,46 @@
+#### Install with Homebrew
+Refer [Installing Python 3 on Mac OS X](https://docs.python-guide.org/starting/install3/osx/)
+
+### Linux
```
Linux should be first since it's our primary supported OS.

gianm commented on docs/operations/single-server.md:
```diff
@@ -44,6 +47,21 @@ The example configurations run the Druid Coordinator and Overlord together in a
+
+`start-druid` is a generic launch script capable of starting any set of Druid services on a server.
+It accepts optional arguments such as list of services, total memory and a config directory to override default jvm arguments and service-specific runtime properties.
+All other reference configurations (e.g. `micro`, `small`, `xlarge`) can be obtained by passing the
```
I would cut this sentence, I think it does not add much other than some confusion.

findingrish replied:
Removed.

gianm commented on docs/operations/single-server.md:
```diff
@@ -44,6 +47,21 @@ The example configurations run the Druid Coordinator and Overlord together in a
+It accepts optional arguments such as list of services, total memory and a config directory to override default jvm arguments and service-specific runtime properties.
+All other reference configurations (e.g. `micro`, `small`, `xlarge`) can be obtained by passing the
+correct memory to this script.
+Druid services will use all processors and upto 80% memory on the system.
```
up to (spelling)

gianm commented on docs/operations/single-server.md:
```diff
@@ -44,6 +47,21 @@ The example configurations run the Druid Coordinator and Overlord together in a
 
+## Druid auto start
+
+Setting up a new Druid cluster can sometimes be complicated as there are several runtime properties required by each service. Each process must also be given the appropriate jvm arguments so that the system can perform optimally.
```
Let's be less apologetic with the wording. I suggest this instead:

> Druid includes a launch script, `bin/start-druid` that automatically sets various memory-related parameters based on available processors and memory. It accepts optional arguments such as list of services, total memory and a config directory to override default jvm arguments and service-specific runtime properties.

gianm commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,32 @@
+then
+  exec python2 "$WHEREAMI/start-druid-main.py" "$@"
+else
+  exec "$WHEREAMI/start-druid-main.py" "$@"
```
This line isn't going to do anything good, because `start-druid-main.py` is not executable. Better to do `exec python "$WHEREAMI/start-druid-main.py" "$@"` in case `python` is the only interpreter available.

findingrish replied:
Updated. I had tested this with Python2 and Python3 interpreter on my machine, missed the `else` part.

vogievetsky commented on examples/bin/start-druid:
```diff
@@ -0,0 +1,32 @@
+then
+  exec python2 "$WHEREAMI/start-druid-main.py" "$@"
+else
+  exec python "$WHEREAMI/start-druid-main.py" "$@"
```
this should do a `elif [ -x "$(command -v python)" ]` and if that fails (in the else) it should print out a nice warning that python could not be found and is needed

vogievetsky replied:
This can be done as a followup also

findingrish replied:
Done.

vogievetsky reviewed (approved):
Amazing

gianm reviewed (approved):
LGTM, thank you! Let's remember, as part of the next release cycle, to test this on a bunch of OSes so we can be sure things work as expected.

vogievetsky commented on docs/operations/single-server.md:
```diff
@@ -23,14 +23,17 @@ title: "Single server deployment"
+Druid includes a set of reference configurations and launch scripts for single-machine deployments.
+These configuration bundles are located in `conf/druid/single-server/`.
+
+The `auto` configuration sizes runtime parameters based on available processors and memory. Other configurations include hard-coded runtime parameters for various server sizes. Most users should stick with `auto`. Refer below [Druid auto start](#Druid-auto-start)
```
I think the anchor needs to be `#druid-auto-start` not `#Druid-auto-start`

vogievetsky commented on examples/bin/supervise:
```diff
@@ -179,13 +195,23 @@ usage() unless GetOptions(
 );
 
-usage() unless $opt{'conf'} && $opt{'vardir'};
+usage() unless ((@{$opt{'command'}} || $opt{'conf'}) && $opt{'vardir'});
```
I don't know enough perl but something in this line is messed up

When I run

```
./bin/start-micro-quickstart 
Can't use an undefined value as an ARRAY reference at /Users/vadim/Projects/apache-druid/distribution/target/apache-druid-26.0.0-SNAPSHOT/bin/supervise line 202.
```

findingrish replied:
Fixed.

gianm commented on examples/bin/supervise:
```diff
@@ -179,13 +195,23 @@ usage() unless GetOptions(
 );
 
-usage() unless $opt{'conf'} && $opt{'vardir'};
+usage() unless (($opt{'conf'} || @{$opt{'command'}}) && $opt{'vardir'});
```
Ah, this should be:

```
usage() unless (($opt{'command'} && @{$opt{'command'}}) || $opt{'conf'}) && $opt{'vardir'}
```

breaking down `($opt{'command'} && @{$opt{'command'}})`:

- `$opt{'command'}` means "there is some value at `$opt{'command'}`
- we expect this to be an arrayref, so, `@{$opt{'command'}}` means "the arrayref at `$opt{'command'}` is nonempty"

findingrish replied:
Done.

maytasm commented:
Did we forgot to change the default values in Configuration reference doc page?
For example, the default value of `druid.cache.sizeInBytes` for caffeine was change from `min(1GiB, Runtime.maxMemory / 10) to `min(1GiB, Runtime.maxMemory / 20)`

findingrish commented:
> Did we forgot to change the default values in Configuration reference doc page? For example, the default value of `druid.cache.sizeInBytes` for caffeine was change from `min(1GiB, Runtime.maxMemory / 10)` to `min(1GiB, Runtime.maxMemory / 20)`

Thanks for pointing it out. I will update it.
